index/index.go

```diff
@@ -0,0 +1,94 @@
+package index
+
+import (
+	"encoding/json"
+	"io"
+	"sync"
+)
+
+type Index interface {
+	Match(terms []string) []string
+	Add(doc string, terms []string)
+	WriteTo(w io.Writer) error
+}
+
+type MapIndex struct {
+	mu    sync.RWMutex
+	Docs  []string
+	Terms map[string][]int
+}
+
+func ReadMapIndex(r io.Reader) (*MapIndex, error) {
+	dec := json.NewDecoder(r)
+	ind := &MapIndex{}
+	if err := dec.Decode(ind); err != nil {
+		return nil, err
+	}
+	return ind, nil
+}
+
+func (i *MapIndex) Match(terms []string) []string {
+	i.mu.RLock()
+	defer i.mu.RUnlock()
+
+	var docsLists [][]int
+	for _, t := range terms {
+		docs := i.Terms[t]
+		if len(docs) > 0 {
+			docsLists = append(docsLists, docs)
+		}
+	}
+	if len(docsLists) == 0 {
+		return nil
```

Comment:
Looks like this could return an empty list below as well. Are callers expected to handle both? Or should this also return an empty list?

Reply:
Empty slices and nil slices behave equivalently in all cases except when being compared to `nil`.

Reply:
In fact, I don't think this function will ever return a non-nil empty list, as the results slice is also nil until a match is found.